Refactor bundle view_url to not instaniate bundle on server components

ephraimbuddy opened the pull request:
This refactor introduces a view_url_template in bundle configuration which is saved in the db and rendered when needed. The url is signed with [api]secret_key for security. Also the view_url_template is also checked for safety before saving to the DB

ashb commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -81,6 +83,54 @@ def _add_example_dag_bundle(config_list):
+    :param bundle_name: The name of the bundle (used in the payload)
+    :return: The signed URL token
+    """
+    serializer = URLSafeSerializer(conf.get_mandatory_value("api", "secret_key"))
```
Why not use Fernet?

ephraimbuddy replied:
I will change to that. It didn't occur to me to use fernet

ashb commented on airflow-core/src/airflow/migrations/versions/0075_3_1_0_add_url_and_template_params_to_.py:
```diff
@@ -0,0 +1,53 @@
+def upgrade():
+    """Apply Add url and template params to DagBundleModel."""
+    with op.batch_alter_table("dag_bundle", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("url", sa.String(length=200), nullable=True))
```
Shouldn't this be "url_template"?

ephraimbuddy replied:
Yeah. Makes sense

ashb commented on providers/git/src/airflow/providers/git/bundles/git.py:
```diff
@@ -247,3 +246,41 @@ def view_url(self, version: str | None = None) -> str | None:
                 return template
         return None
+
+    def view_url_template(self) -> str | None:
```
The now deprecated `view_url` is very similar to this method. Is it worth refactoring it so that `view_url` calls this fn and then renders the template?

ephraimbuddy replied:
Ok. Will do that. Problem was the view_url accepting a version but I think I can work around it

ephraimbuddy replied:
There's a little difference that won't allow us refactor this: the view_url_template returns a template while view_url doesn't return a template but full url. Changing it to return a template would be a breaking change

ashb replied:
Can view url use the template and render it?

ephraimbuddy replied:
Yeah. was possible. Thanks so much. https://github.com/apache/airflow/pull/52876/commits/c6561d19f22aeda267c1c6fa52a62934ee4a4fa6

ephraimbuddy committed:
fixup! Refactor bundle view_url to not instaniate bundle on server components

ephraimbuddy committed:
Rename bundle url to url_template and fix backcompat

ephraimbuddy committed:
Refactor render_url to not depend on dagbundlemodel version

ephraimbuddy committed:
Render view_url_template in view_url

ephraimbuddy committed:
fixup! Render view_url_template in view_url

ephraimbuddy committed:
Add deprecation warning and update s3 bundle with view_url_template

ephraimbuddy committed:
fixup! Add deprecation warning and update s3 bundle with view_url_template

ephraimbuddy committed:
Remove deprecation warning in provider's view_url

pierrejeambrun reviewed (commented):
LGTM, but we probably want another pair of eyes as this PR is quite big. (easy to miss something)

ephraimbuddy committed:
fixup! Remove deprecation warning in provider's view_url

ephraimbuddy committed:
fixup! fixup! Remove deprecation warning in provider's view_url

ephraimbuddy committed:
Add backcompat for bundles vs airflow releases

ephraimbuddy committed:
Refactor bundle view_url to not instaniate bundle on server components

This refactor introduces a view_url_template in bundle configuration
which is saved in the db and rendered when needed. The url is signed
with [api]secret_key for security. Also the view_url_template is
also checked for safety before saving to the DB

ephraimbuddy committed:
fixup! Add backcompat for bundles vs airflow releases

amoghrajesh commented on airflow-core/src/airflow/models/dagbundle.py:
```diff
@@ -40,6 +44,59 @@ class DagBundleModel(Base):
     active = Column(Boolean, default=True)
     version = Column(String(200), nullable=True)
     last_refreshed = Column(UtcDateTime, nullable=True)
+    url_template = Column(String(200), nullable=True)
```
nit: `signed_url_template` to be more explicit about what's stored?

ephraimbuddy replied:
Yeah. Makes sense

amoghrajesh commented on airflow-core/src/airflow/models/dagbundle.py:
```diff
@@ -40,6 +44,59 @@ class DagBundleModel(Base):
+        url_template = self._unsign_url()
+
+        if url_template is None:
+            url_template = self.url_template
```
If we fail to unsign, we get the signed one? Shouldn't we fallback to some default?

ephraimbuddy replied:
Would None be better than returning the signed url? or do you have a default in mind?

amoghrajesh replied:
What behaviour did we have earlier? If we couldn't construct the url, would we return null?

ephraimbuddy replied:
I guess we should go with None. Yeah, previously we would return None

ephraimbuddy replied:
Fixed

amoghrajesh commented on providers/amazon/src/airflow/providers/amazon/aws/bundles/s3.py:
```diff
@@ -137,10 +137,20 @@ def refresh(self) -> None:
-
+        if hasattr(self, "_view_url_template") and self._view_url_template:
+            # Backward compatibility for released Airflow versions
+            return self._view_url_template
```
I do not understand this compat. The `_view_url_template` is a new field right? How is it present on older Airflow?

ephraimbuddy replied:
Airflow 3.0.2 for example doesn't have the `_view_url_template`. If this provider is released and used in `3.0.2`, it will raise errors

amoghrajesh replied:
Alright, i see that this is to achieve compat between older airflow and new provider.

It makes sense, thanks!

amoghrajesh commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -125,11 +175,56 @@ def parse_config(self) -> None:
+                    else:
+                        # Sign the URL for integrity verification
+                        new_template = _sign_bundle_url(new_template, name)
+                        self.log.debug("Signed URL template for bundle %s", name)
```
Can we extract this part into a function, seems its repeated below too?

amoghrajesh reviewed (commented):
The usage pattern looks good, i have some qns around compat

ephraimbuddy committed:
rename url_template to signed_url_template. Also return None when we can't unsign a url

ephraimbuddy committed:
refactor template signing

ephraimbuddy committed:
fixup! refactor template signing

jedcunningham commented on airflow-core/src/airflow/dag_processing/bundles/base.py:
```diff
@@ -316,10 +323,31 @@ def view_url(self, version: str | None = None) -> str | None:
         """
+        warnings.warn(
+            "The 'view_url' method is deprecated and will be removed in a future version. "
+            "Use 'view_url_template' instead.",
```
Should we point folks at `render_url` instead of in addition to `view_url_template`?

ephraimbuddy replied:
The `render_url` is `DagBundleModel`'s method and not `BaseDagBundle/DagBundle` method. The deprecation on Bundle manager suggested render_url which is more appropriate there

jedcunningham commented on providers/amazon/src/airflow/providers/amazon/aws/bundles/s3.py:
```diff
@@ -137,10 +137,20 @@ def refresh(self) -> None:
+        """
+        Return a URL for viewing the DAGs in S3. Currently, versioning is not supported.
+
+        This method is deprecated and will be removed in a future release. Use `view_url_template` instead.
```
We should leave a breadcrumb as to when we can remove it, e.g. when we drop support for 3.0.

Maybe we conditionally add a deprecation warning based on the core version?

ephraimbuddy replied:
Yeah. I added deprecation warning before and it a lot of `pytest.warns(AirflowProviderDeprecationWarning)` in core tests/providers tests and I decided to remove it. Here's the commits: https://github.com/apache/airflow/pull/52876/commits/74c3eb8e214db8faf83958a9c89ed38656548f3b and https://github.com/apache/airflow/pull/52876/commits/74c3eb8e214db8faf83958a9c89ed38656548f3b. Should I add it back or just leave a comment about the deprecation version

jedcunningham commented on providers/amazon/src/airflow/providers/amazon/aws/bundles/s3.py:
```diff
@@ -137,10 +137,20 @@ def refresh(self) -> None:
             )
 
     def view_url(self, version: str | None = None) -> str | None:
```
Is this used for non-versioned bundles? That feels like a bug...

ephraimbuddy replied:
It returns None for non-versioned bundles which I think is ok

jedcunningham commented on providers/git/src/airflow/providers/git/bundles/git.py:
```diff
@@ -218,32 +217,53 @@ def _convert_git_ssh_url_to_https(url: str) -> str:
         for allowed_host, template in host_patterns.items():
             if host == allowed_host or host.endswith(f".{allowed_host}"):
+                if self.subdir:
+                    return f"{template}/{self.subdir}"
```
We don't seem to be using the whole `template_fields` and `params` complexity, and just adding subdir straight on here. Do we need that concept at all? I don't see why the template can't just have that baked in.

I could see supporting more things down the road (e.g. relative path), but this template_fields / params would be static per bundle anyways.

ephraimbuddy replied:
My bad. Fixing this

jedcunningham replied:
Unresolving this. It's basically the same original concern, just in a more roundabout manner now - it's just shoved on the end.

ephraimbuddy replied:
We could get wrong information if the version is resolved here. Because, the version changes and what's stored in the dagbundlemodel.version also changes which means the url will keep changing. What I did here was to resolve the url with dag_version.bundle version information when accessed from the API. That's why `render_url` is not using the version on the DagBundleModel but uses the supplied version which comes from dag_version.bundle_version

jedcunningham replied:
It's not version - that we need to interpolate at render time as it varies. I mean how we are handling subdir, and needing to add the whole template fields concept to accommodate it. I'm suggesting we don't need any of that.

ephraimbuddy replied:
The template_fields is also for the view_template_url configuration. It is not explicit but if a user decides to provide a custom view_url_template at configuration, the template_fields is used to render it:
```config
{
                "name": "my-git-repo",
                "classpath": "airflow.providers.git.bundles.git.GitDagBundle",
                "kwargs": {
                  "subdir": "dags",
                  "tracking_ref": "main",
                  "refresh_interval": 0,
                  "view_template_url":"{base_url}/{version}/{some_fields}"
                },
                
              }
```
This was the idea but I think it's not complete. The other kwargs should extend the template_fields. Let me know if these should be discarded

jedcunningham replied:
Okay, I do see value in that so folks don't need to subclass bundles for the view_template_url use case.

I do wonder if we should use a different name than `templated_fields`, just because its the same name as baseop, but is the opposite where those files are used for templating but not templated themselves.

Also worth commenting on what it is in the base class too.

jedcunningham commented on providers/git/src/airflow/providers/git/bundles/git.py:
```diff
@@ -218,32 +217,53 @@ def _convert_git_ssh_url_to_https(url: str) -> str:
+
+    def view_url_template(self) -> str | None:
+        if hasattr(self, "_view_url_template") and self._view_url_template:
+            # Backward compatibility for released Airflow versions
```
This comment is a bit misleading, this is either backcompat or when `view_url_template` isn't configured manually right?

ephraimbuddy replied:
Not really. The released airflow versions don't have `_view_url_template` attr so using this provider with them will result in attribute error.

amoghrajesh replied:
Maybe you can explain it better in terms of usage for older airflow and new provider, maybe update the comment to along the lines of:

```
This is to achieve back compat. Older airflow doesn't have this property set, so if this is present, it indicates airflow version that can render view url using template string, otherwise it uses the older mechanism. 
```

Give or take..

jedcunningham commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -124,12 +174,55 @@ def parse_config(self) -> None:
+                # Update URL template and parameters if they've changed
+                bundle_instance = self.get_bundle(name)
+                new_template = bundle_instance.view_url_template()
+                new_params = self._extract_template_params(bundle_instance)
```
We should dedup what we can between this if/else block.

ephraimbuddy committed:
Fix templating and conflict

amoghrajesh commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/dag_versions.py:
```diff
@@ -41,10 +42,21 @@ class DagVersionResponse(BaseModel):
+
+                if bundle_model and hasattr(bundle_model, "signed_url_template"):
+                    return bundle_model.render_url(self.bundle_version)
+                try:
+                    return DagBundlesManager().view_url(self.bundle_name, self.bundle_version)
+                except ValueError:
+                    return None
```
Now that I think of this, try catch is probably not needed because the `view_url` would handle that.

ephraimbuddy replied:
Nope. It doesn't handle it. The ValueError comes from unconfigured bundle or removed bundle

amoghrajesh replied:
Makes sense, thanks!

amoghrajesh commented on airflow-core/src/airflow/dag_processing/bundles/base.py:
```diff
@@ -218,6 +218,9 @@ def remove_stale_bundle_versions(self):
         log.info("checking for stale bundle versions locally")
+
+        from airflow.dag_processing.bundles.manager import DagBundlesManager
+
```
nit: import above log?

amoghrajesh commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -124,12 +174,45 @@ def parse_config(self) -> None:
+            # Update URL template and parameters if they've changed
+            bundle_instance = self.get_bundle(name)
+            new_template = bundle_instance.view_url_template()
+            new_params = self._extract_template_params(bundle_instance)
```
I wonder if we can just move extraction of these two directly into the signed_template and call it:
`extract_and_sign_bundle_template(bundle_instance)` or even better move line 197 too. That way, it only takes bundle_name.

amoghrajesh reviewed (commented):
@ephraimbuddy looking much better, some more nits.

I think we are looking close!

ephraimbuddy committed:
Fix backcompat & refactor template signing

ephraimbuddy commented on providers/amazon/tests/unit/amazon/aws/bundles/test_s3.py:
```diff
@@ -109,9 +109,18 @@ def test_view_url_generates_presigned_url(self):
         url: str = bundle.view_url("test_version")
         assert url.startswith("https://my-airflow-dags-bucket.s3.amazonaws.com/project1/dags")
 
+    @pytest.mark.db_test
```
```suggestion
```

ephraimbuddy replied:
Needed

ephraimbuddy commented on airflow-core/src/airflow/dag_processing/bundles/base.py:
```diff
@@ -316,10 +323,34 @@ def view_url(self, version: str | None = None) -> str | None:
         """
+        warnings.warn(
+            "The 'view_url' method is deprecated and will be removed in a future version. "
+            "Use 'view_url_template' instead.",
```
```suggestion
            "Use 'DagBundleModel.render_url' instead.",
```

ephraimbuddy committed:
fixup! Fix backcompat & refactor template signing

ephraimbuddy committed:
fixup! fixup! Fix backcompat & refactor template signing

ephraimbuddy committed:
skip some test if not airflow 3.1+

ephraimbuddy committed:
fixup! skip some test if not airflow 3.1+

amoghrajesh commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/dag_versions.py:
```diff
@@ -41,10 +42,21 @@ class DagVersionResponse(BaseModel):
+                if bundle_model and hasattr(bundle_model, "signed_url_template"):
+                    return bundle_model.render_url(self.bundle_version)
+                try:
+                    return DagBundlesManager().view_url(self.bundle_name, self.bundle_version)
```
Can we add a comment here please?

amoghrajesh commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -124,12 +174,44 @@ def parse_config(self) -> None:
     def sync_bundles_to_db(self, *, session: Session = NEW_SESSION) -> None:
         self.log.debug("Syncing DAG bundles to the database")
+
+        def _extract_and_sign_template(bundle_name: str) -> tuple[str | None, dict]:
+            bundle_instance = self.get_bundle(name)
+            new_template_ = bundle_instance.view_url_template()
+            new_params_ = self._extract_template_params(bundle_instance)
+            if new_template_:
+                if not _is_safe_bundle_url(new_template_):
+                    self.log.warning(
+                        "Bundle %s has unsafe URL template '%s', skipping URL update",
+                        bundle_name,
+                        new_template_,
+                    )
+                    new_template_ = None
+                else:
+                    # Sign the URL for integrity verification
+                    new_template_ = _sign_bundle_url(new_template_, bundle_name)
+                    self.log.debug("Signed URL template for bundle %s", bundle_name)
+            return new_template_, new_params_
```
Really nice!

amoghrajesh commented on airflow-core/src/airflow/models/dagbundle.py:
```diff
@@ -32,6 +34,8 @@ class DagBundleModel(Base):
     - version: The latest version Airflow has seen for the bundle.
     - last_refreshed: When the bundle was last refreshed.
+    - signed_url_template: Signed URL template for viewing the bundle
+    - template_params: JSON object containing template parameters (e.g., {"subdir": "dags"})
```
```suggestion
    - template_params: JSON object containing template parameters for constructing view url (e.g., {"subdir": "dags"})
```

amoghrajesh commented on airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dags.py:
```diff
@@ -507,8 +507,10 @@ class TestDagDetails(TestDagEndpoint):
         ],
     )
     @pytest.mark.usefixtures("configure_git_connection_for_dag_bundle")
+    @mock.patch("airflow.api_fastapi.core_api.datamodels.dag_versions.hasattr")
```
Non blocking, but create a fixture here?
(skip if you think it doesn't make sense)

ephraimbuddy replied:
Skipping. These are temporary tests that would be removed once provider minimum supported version is 3.1

amoghrajesh reviewed (approved):
Looking good to me now, @jedcunningham could you look at it from a more dag bundles side of things?

ephraimbuddy committed:
fixup! fixup! skip some test if not airflow 3.1+

ephraimbuddy committed:
Resolve conflict

ephraimbuddy committed:
Add version to be removed for deprecated view_url

ephraimbuddy committed:
Remove template_fields and use regex to extract placeholders

ephraimbuddy committed:
Fix conflict

ephraimbuddy committed:
fixup! Remove template_fields and use regex to extract placeholders

ephraimbuddy committed:
Remove added deadline in dag details

uranusjr commented on airflow-core/docs/administration-and-deployment/dag-bundles.rst:
```diff
@@ -83,6 +83,27 @@ For example, adding multiple dag bundles to your ``airflow.cfg`` file:
+      ]
+
+Above, the ``view_url_template`` is set to a custom URL that will be used to view the Dags in the ``my_git_repo`` bundle. The ``{subdir}`` placeholder will be replaced
+with the ``subdir`` attribute of the bundle. When you specify a custom URL, it overrides the default URL provided by the dag bundle.
```
What placeholders can the user use in the template? This should be listed/described in the documentation.

uranusjr commented on airflow-core/src/airflow/dag_processing/bundles/base.py:
```diff
@@ -316,10 +322,31 @@ def view_url(self, version: str | None = None) -> str | None:
+
+    def view_url_template(self) -> str | None:
+        """
+        URL template to view the bundle on an external website. This is shown to users in the Airflow UI, allowing them to navigate to this url for more details about that version of the bundle.
```
```suggestion
        URL template to view the bundle on an external website.
        
        This is shown to users in the Airflow UI, allowing them to navigate to this url for more details about that version of the bundle.
```

uranusjr commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -81,6 +83,54 @@ def _add_example_dag_bundle(config_list):
     )
 
 
+def _is_safe_bundle_url(url: str) -> bool:
```
The documentation should mention these restrictions.

ephraimbuddy replied:
Done

uranusjr commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -81,6 +83,54 @@ def _add_example_dag_bundle(config_list):
+            return False
+
+        if ";" in url:
+            return False
```
Why?

ephraimbuddy replied:
I was thinking it could lead to path manipulation attack. WDYT?

uranusjr replied:
Semicolon is a valid character in a URL and has special meaning in query strings, although not used much these days (generally `&` is preferred) so banning it could break some things. I’m inclined to not treat it specially since it’s a user config (i.e. modifying it requires admin permission).

Honestly I’m inclined to not do any validation at all instead since your system is already fried if someone can put malicious content in this value. But some sanity checks (like checking for control characters) is not a bad idea too.

ephraimbuddy replied:
Got it. Take a look again on the PR, only now checks if it's http/https and if it has a control character

ephraimbuddy committed:
Update docs

uranusjr commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -81,6 +83,54 @@ def _add_example_dag_bundle(config_list):
+        if ";" in url:
+            return False
+        if any(ord(c) < 32 for c in url):
+            return False
```
It may be a better idea for this function to raise an exception on invalid inputs instead (and caught when called). This way the warning message can contain more information on exactly what’s wrong with the user-provided value.

ephraimbuddy replied:
Looking at this. I think we can log instead of raising exception so we don't block other bundles from parsing

ephraimbuddy committed:
Remove ; from url safety check

ephraimbuddy committed:
Log url sanity check errors

jedcunningham commented on airflow-core/tests/unit/cli/commands/test_asset_command.py:
Why is this being added? Doesn't seem related.

ephraimbuddy replied:
I added a test on this file and updated one too. I had to keep both since we haven't removed the deprecated method. So this acts like a check that there's compatibility

jedcunningham commented on providers/amazon/tests/unit/amazon/aws/bundles/test_s3.py:
```diff
@@ -109,9 +109,18 @@ def test_view_url_generates_presigned_url(self):
         assert url.startswith("https://my-airflow-dags-bucket.s3.amazonaws.com/project1/dags")
 
+    @pytest.mark.db_test
+    def test_view_url_template_generates_presigned_url(self):
```
Not sure this test name is right, it generates a url, but not a presigned one?

ephraimbuddy replied:
```suggestion
    def test_view_url_template_generates_url(self):
```

ephraimbuddy commented on airflow-core/src/airflow/dag_processing/bundles/manager.py:
```diff
@@ -140,6 +229,35 @@ def sync_bundles_to_db(self, *, session: Session = NEW_SESSION) -> None:
+        # Extract template placeholders using regex
+        # This matches {placeholder} patterns in the template
+        placeholder_pattern = r"\{([^}]+)\}"
+        placeholders = re.findall(placeholder_pattern, template)
```
@uranusjr what do you think about this use of regex here?

ephraimbuddy committed:
compile re
